File: Impl/src/main/java/rock/clanattack/java/AnnotationScanner.java
package rock.clanattack.java;

import io.github.classgraph.ClassGraph;
import rocks.clanattack.entry.plugin.Loader;
import rocks.clanattack.impl.util.JavaMap;

import java.lang.annotation.Annotation;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

public class AnnotationScanner {

    public static List<Class<?>> getAnnotatedClasses(Class<? extends Annotation> clazz, ClassLoader loader, String basePackage) {
        try (var result = new ClassGraph()
                .enableClassInfo()
                .enableAnnotationInfo()
                .addClassLoader(loader)
                .acceptPackages(basePackage)
                .scan()) {
            return result.getClassesWithAnnotation(clazz)
                    .loadClasses();
        }
    }

    public static List<Class<?>> getAnnotatedClasses(Class<? extends Annotation> clazz) {
        var loaders = JavaMap.find(Loader.class).getClassLoaders().entrySet();
        var glassGraph = new ClassGraph()
                .enableClassInfo()
                .enableAnnotationInfo()
                .acceptPackages(loaders.stream().map(Map.Entry::getValue).toArray(String[]::new));

        for (var classLoader : loaders) glassGraph.addClassLoader(classLoader.getKey());


        try (var result = glassGraph.scan()) {
            return result.getClassesWithAnnotation(clazz).loadClasses();
        }
    }

    public static List<Method> getAnnotatedMethods(Class<? extends Annotation> clazz, ClassLoader loader, String basePackage) {
        try (var result = new ClassGraph()
                .enableClassInfo()
                .enableMethodInfo()
                .enableAnnotationInfo()
                .addClassLoader(loader)
                .acceptPackages(basePackage)
                .scan()) {

            return result.getClassesWithMethodAnnotation(clazz)
                    .loadClasses()
                    .stream()
                    .flatMap(c -> Arrays.stream(c.getMethods()))
                    .filter(m -> m.isAnnotationPresent(clazz))
                    .toList();

        }
    }

    public static List<Method> getAnnotatedMethods(Class<? extends Annotation> clazz) {
        var loaders = JavaMap.find(Loader.class).getClassLoaders().entrySet();
        var glassGraph = new ClassGraph()
                .enableClassInfo()
                .enableMethodInfo()
                .enableAnnotationInfo()
                .acceptPackages(loaders.stream().map(Map.Entry::getValue).toArray(String[]::new));

        for (var classLoader : loaders) glassGraph.addClassLoader(classLoader.getKey());

        try (var result = glassGraph.scan()) {
            return result.getClassesWithMethodAnnotation(clazz)
                    .loadClasses()
                    .stream()
                    .flatMap(c -> Arrays.stream(c.getMethods()))
                    .filter(m -> m.isAnnotationPresent(clazz))
                    .toList();
        }
    }

}

File: Impl/src/main/kotlin/rocks/clanattack/impl/util/annotation/AnnotationScanner.kt
package rocks.clanattack.impl.util.annotation

import org.reflections.Reflections
import org.reflections.scanners.MethodAnnotationsScanner
import org.reflections.scanners.SubTypesScanner
import org.reflections.scanners.TypeAnnotationsScanner
import org.reflections.util.ClasspathHelper
import org.reflections.util.ConfigurationBuilder
import rocks.clanattack.entry.find
import rocks.clanattack.entry.plugin.Loader
import kotlin.reflect.KClass

object AnnotationScanner {

    fun findClasses(annotation: KClass<out Annotation>, classLoader: ClassLoader, basePackage: String) =
        Reflections(
            ConfigurationBuilder()
                .addClassLoader(classLoader)
                .setUrls(ClasspathHelper.forPackage(basePackage, classLoader))
                .setScanners(TypeAnnotationsScanner(), SubTypesScanner())
        ).getTypesAnnotatedWith(annotation.java)
            .map { it.kotlin }

    fun findClasses(annotation: KClass<out Annotation>) =
        Reflections(
            ConfigurationBuilder()
                .addClassLoaders(find<Loader>().classLoaders.keys)
                .setUrls(find<Loader>().classLoaders.map { (classLoader, basePackage) ->
                    ClasspathHelper.forPackage(
                        basePackage,
                        classLoader
                    )
                }.flatten())
                .setScanners(TypeAnnotationsScanner(), SubTypesScanner())
        ).getTypesAnnotatedWith(annotation.java)
            .map { it.kotlin }

    fun findMethods(annotation: KClass<out Annotation>, classLoader: ClassLoader, basePackage: String) =
        Reflections(
            ConfigurationBuilder()
                .addClassLoader(classLoader)
                .setUrls(ClasspathHelper.forPackage(basePackage, classLoader))
                .setScanners(MethodAnnotationsScanner(), SubTypesScanner())
        ).getMethodsAnnotatedWith(annotation.java)
            .filterNotNull()

    fun findMethods(annotation: KClass<out Annotation>) =
        Reflections(
            ConfigurationBuilder()
                .addClassLoaders(find<Loader>().classLoaders.keys)
                .setUrls(find<Loader>().classLoaders.map { (classLoader, basePackage) ->
                    ClasspathHelper.forPackage(
                        basePackage,
                        classLoader
                    )
                }.flatten())
                .setScanners(MethodAnnotationsScanner(), SubTypesScanner())
        ).getMethodsAnnotatedWith(annotation.java)
            .filterNotNull()


}